File: src/api/java/org/models/organization/relation/manager/NeedsManager.java
package org.models.organization.relation.manager;

import java.util.Set;

import org.models.organization.entity.Attribute;
import org.models.organization.entity.Role;
import org.models.organization.id.UniqueId;
import org.models.organization.relation.Needs;

/**
 * The {@linkplain NeedsManager} interface defines the API for managing {@linkplain Needs}.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
public interface NeedsManager {
	/**
	 * Creates a needs relation between a {@linkplain Role} and an {@linkplain Attribute} in this {@linkplain NeedsManager}.
	 *
	 * @param roleId
	 *            the {@linkplain UniqueId} that represents the {@linkplain Role}.
	 * @param attributeId
	 *            the {@linkplain UniqueId} that represents the {@linkplain Attribute}.
	 */
	void addNeeds(UniqueId<Role> roleId, UniqueId<Attribute> attributeId);

	/**
	 * Returns a set of {@linkplain Attribute}s that is needed by a {@linkplain Role} in this {@linkplain NeedsManager}.
	 *
	 * @param id
	 *            the {@linkplain UniqueId} that represents the {@linkplain Role}.
	 * @return a set of {@linkplain Attribute}s.
	 */
	Set<Attribute> getNeeds(UniqueId<Role> id);

	/**
	 * Returns a set of {@linkplain Role}s that needs an {@linkplain Attribute} in this {@linkplain NeedsManager}.
	 *
	 * @param id
	 *            the {@linkplain UniqueId} that represents the {@linkplain Attribute}.
	 * @return a set of {@linkplain Role}s.
	 */
	Set<Role> getNeededBy(UniqueId<Attribute> id);

	/**
	 * Removes a needs relation between a {@linkplain Role} and an {@linkplain Attribute} from this {@linkplain NeedsManager}.
	 *
	 * @param roleId
	 *            the {@linkplain UniqueId} that represents the {@linkplain Role}.
	 * @param attributeId
	 *            the {@linkplain UniqueId} that represents the {@linkplain Attribute}.
	 */
	void removeNeeds(UniqueId<Role> roleId, UniqueId<Attribute> attributeId);

	/**
	 * Removes all needs relations from this {@linkplain NeedsManager}.
	 */
	void removeAllNeeds();

}

File: chazm-api/src/main/kotlin/runtimemodels/chazm/api/organization/NeedsManager.kt
package runtimemodels.chazm.api.organization

import runtimemodels.chazm.api.entity.Attribute
import runtimemodels.chazm.api.entity.Role
import runtimemodels.chazm.api.id.AttributeId
import runtimemodels.chazm.api.id.RoleId
import runtimemodels.chazm.api.relation.Needs

/**
 * The [NeedsManager] interface defines the APIs for managing a set of [Needs] relations.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
interface NeedsManager {
    /**
     * Adds a [Needs] relation between a [Role] and an [Attribute] to this [NeedsManager].
     *
     * @param needs the [Needs] relation to add.
     */
    fun add(needs: Needs)

    /**
     * Returns the [Needs] relation between an [Role] and an [Attribute] from this [HasManager].
     *
     * @param roleId           the [RoleId] that represents the [Role].
     * @param attributeId the [AttributeId] that represents the [Attribute].
     * @return the [Needs] relation if it exists, `null` otherwise.
     */
    operator fun get(roleId: RoleId, attributeId: AttributeId): Needs?

    /**
     * Returns a [Map] of [Attribute]s that are contained by a [Role] from this [NeedsManager].
     *
     * @param id the [RoleId] that represents the [Role].
     * @return a [Map] of [Attribute]s.
     */
    operator fun get(id: RoleId): Map<AttributeId, Needs>

    /**
     * Returns a [Map] of [Role]s that contains a [Attribute] from this [NeedsManager].
     *
     * @param id the [AttributeId] that represents the [Attribute].
     * @return a [Map] of [Role]s.
     */
    operator fun get(id: AttributeId): Map<RoleId, Needs>

    /**
     * Removes a [Needs] relation between a [Role] and an [Attribute] from this [NeedsManager].
     *
     * @param roleId      the [RoleId] that represents the [Role].
     * @param attributeId the [AttributeId] that represents the [Attribute].
     * @return the [Needs] relation that was removed, `null` otherwise.
     */
    fun remove(roleId: RoleId, attributeId: AttributeId): Needs?

    /**
     * Removes all [Needs] relations associated to a [Role] from this [NeedsManager].
     *
     * @param id the [RoleId] that represents the [Role].
     */
    fun remove(id: RoleId)

    /**
     * Removes all [Needs] relations associated to a [Attribute] from this [NeedsManager].
     *
     * @param id the [AttributeId] that represents the [Attribute].
     */
    fun remove(id: AttributeId)
}
